src/local.rs

```diff
@@ -105,29 +92,45 @@ impl Local {
         )
         .context(OpenDatabaseSnafu {})?;
 
-        // Get the relative directory of the maildir to the database path.
-        let canonical_db_path = db.path().canonicalize().context(CanonicalizeSnafu {})?;
-        let canonical_mail_dir_path = mail_dir
-            .as_ref()
+        // Get notmuch's idea of the mail root. If, for whatever reason, we get nothing back for
+        // that key (ancient version of notmuch?), use the database path.
+        let mail_root = db
+            .config(ConfigKey::MailRoot)
+            .map_or(
+                db.path().into(),
+                |root| PathBuf::from(root)
```

Comment:
Probably don't need to clone into owned types here since we call `canonicalize()` on the result.

Reply:
`db.path()` returns `&Path`, `root` is a `&str`. An owned `PathBuf` is the only way I can see to make it all live long enough using this particular construction.

Cheap alternative; move `canonicalize` into both arms:

```diff
diff --git src/local.rs src/local.rs
index b347283..e40d2cc 100644
--- src/local.rs
+++ src/local.rs
@@ -15,7 +15,7 @@ use std::collections::HashMap;
 use std::collections::HashSet;
 use std::fs;
 use std::io;
-use std::path::PathBuf;
+use std::path::{Path, PathBuf};
 
 const ID_PATTERN: &'static str = r"[-A-Za-z0-9_]+";
 const MAIL_PATTERN: &'static str = formatcp!(r"^({})\.({})(?:$|:)", ID_PATTERN, ID_PATTERN);
@@ -95,8 +95,9 @@ impl Local {
         // that key (ancient version of notmuch?), use the database path.
         let mail_root = db
             .config(ConfigKey::MailRoot)
-            .map_or(db.path().into(), |root| PathBuf::from(root))
-            .canonicalize()
+            .map_or(db.path().canonicalize(), |root| {
+                Path::new(&root).canonicalize()
+            })
             .context(CanonicalizeSnafu {})?;
 
         // Figure out our maildir. Either the configured thing, or notmuch's mail root. Which in
```